flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,188 @@
+		sicPlanNode.setParallelism(1);
+		sicChannel.setTarget(sicPlanNode);
+		sipPlanNode.addOutgoingChannel(sicChannel);
+		this.plan.getAllNodes().add(sicPlanNode);
+
+		// 3. Use sampled data to build range boundaries.
+		RangeBoundaryBuilder rangeBoundaryBuilder = new RangeBoundaryBuilder(comparator, targetParallelism);
+		TypeInformation<CommonRangeBoundaries> rbTypeInformation = TypeExtractor.getForClass(CommonRangeBoundaries.class);
+		UnaryOperatorInformation rbOperatorInformation = new UnaryOperatorInformation(sourceOutputType, rbTypeInformation);
+		MapPartitionOperatorBase rbOperatorBase = new MapPartitionOperatorBase(rangeBoundaryBuilder, rbOperatorInformation, "RangeBoundaryBuilder");
+		MapPartitionNode rbNode= new MapPartitionNode(rbOperatorBase);
+		Channel rbChannel = new Channel(sicPlanNode, TempMode.NONE);
+		rbChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode rbPlanNode = new SingleInputPlanNode(rbNode, "RangeBoundary PlanNode", rbChannel, DriverStrategy.MAP_PARTITION);
+		rbPlanNode.setParallelism(1);
+		rbChannel.setTarget(rbPlanNode);
+		sicPlanNode.addOutgoingChannel(rbChannel);
+		this.plan.getAllNodes().add(rbPlanNode);
+
+		// 4. Take range boundaries as broadcast input and take the tuple of partition id and record as output.
+		AssignRangeIndex assignRangeIndex = new AssignRangeIndex(comparator);
+		TypeInformation<Tuple2> ariOutputTypeInformation = new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, sourceOutputType);
+		UnaryOperatorInformation ariOperatorInformation = new UnaryOperatorInformation(sourceOutputType, ariOutputTypeInformation);
+		MapPartitionOperatorBase ariOperatorBase = new MapPartitionOperatorBase(assignRangeIndex, ariOperatorInformation, "Assign Range Index");
+		MapPartitionNode ariNode= new MapPartitionNode(ariOperatorBase);
+		Channel ariChannel = new Channel(sourceNode, TempMode.NONE);
+		ariChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode ariPlanNode = new SingleInputPlanNode(ariNode, "AssignRangeIndex PlanNode", ariChannel, DriverStrategy.MAP_PARTITION);
+		ariPlanNode.setParallelism(sourceParallelism);
+		ariChannel.setTarget(ariPlanNode);
+		appendOutgoingChannels.add(ariChannel);
+		this.plan.getAllNodes().add(ariPlanNode);
+
+		channel.setSource(ariPlanNode);
+		ariPlanNode.addOutgoingChannel(channel);
+
+		NamedChannel broadcastChannel = new NamedChannel("RangeBoundaries", rbPlanNode);
+		broadcastChannel.setShipStrategy(ShipStrategyType.BROADCAST, channel.getDataExchangeMode());
+		broadcastChannel.setTarget(ariPlanNode);
+		List<NamedChannel> broadcastChannels = new ArrayList<>(1);
+		broadcastChannels.add(broadcastChannel);
+		ariPlanNode.setBroadcastInputs(broadcastChannels);
+
+		// 5. Remove the partition id.
+		PartitionIDRemoveWrapper partitionIDRemoveWrapper = new PartitionIDRemoveWrapper();
```

Comment:
Ah, yes. You are right! I for got about the Partition node. However, the Partition node operates with a `NoOpDriver` which means that it does nothing. Hence, I think we can safely replace this node by the `PartitionIDRemoveWrapper` which removes the ID after the partitioning instead of nothing. This will avoid some serialization overhead induced by the additional operator.